### PR TITLE
Fixed and improved the docker installation process

### PR DESCRIPTION
**Pull request by nickdbush**
In trying to resolve issues #3113, #3104, #3067, #2930, #2928, #2832 and #2733, I've tried to rework the docker installation process to smooth out some bugs. As it stands, it is fundamentally impossible to orchestrate one's own version of superdesk without significant experience with dockers and containers in general. The biggest changes include:

1. Updating ubuntu and node to their latest versions, as the current versions are outdated. Node would refuse to install, as v7.x is no longer supported. Mongo, Redis and Elastic all seem to be using outdated versions as well. This **needs to be discussed** before merging, as @petrjasek has a elastic7 branch which does better. (8a223846d and 4603aab98)
2. The python start script was being interpreted as a shell script, causing the installation to fail. This was caused by the hash-bang referencing a local python interpreter of the developer. (e28e11513)
3. Versions of the ```docker-compose.yml``` scripts were extending services not defined in ```common.yml```. Logstash has been temporarily removed, as there is ```Dockerfile``` for that service in its subdirectory. ```pubapi``` seems to no longer be a service, so that reference was removed. (0e3aa45ea, f45119bd2 and a8201c97c)
4. There were significant problems around getting the highstash export server to install and run. Therefore, I would suggest disabling the analytics until this is fixed, so people can still install superdesk. Logstash also refuses to connect to elastic, so that service has been temporarily disabled as well. (ccfc98219 and f45119bd2)
6. General formatting of the docker files for consistency. (1a1661a2f and spread throughout other commits)

Looking forward to ironing out these installation issues and working towards an easier onboarding process as more people hopefully move towards superdesk and the open source newsroom dream.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -44,11 +46,12 @@ ENV CELERYBEAT_SCHEDULE_FILENAME /tmp/celerybeatschedule.db
 
 # install server
 COPY ./server /opt/superdesk
-RUN pip3 install -U -r requirements.txt
+RUN python3 -m pip install --upgrade pip setuptools wheel
+RUN python3 -m pip install -U -r requirements.txt --ignore-installed
 
 # install client
 COPY ./client /opt/superdesk/client/
-RUN npm install -g n npm grunt-cli && n lts
```
Comment (nickdbush) [inline Dockerfile]: It felt redundant to install nodejs only to use it to install nodejs again

**Comment by kottkrig** [comment]
I'm getting an error when trying to start the containers from the compose file.

## Steps to reproduce:

Clone the repo

```bash
git clone git@github.com:nickdbush/superdesk.git
```

Checkout docker branch

```bash
cd superdesk
git checkout docker
```

Start the containers

```
cd docker
DOCKER_IP=localhost docker-compose up
```

## Error:

```
> @ compile /opt/superdesk/client/extensions/helloWorld2
> tsc

src/extension.ts(3,7): error TS2741: Property 'id' is missing in type '{ activate: (superdesk: DeepReadonlyObject<{ dataApi: IDataApi; dataApiByEntity: { article: { query(parameters: Pick<IQueryElasticParameters, "page" | "sort" | "filterValues" | "aggregations">): Promise<IArticleQueryResult>; }; }; ... 13 more ...; removeEventListener<T extends "articleEditStart" | "articleEditEnd">(...' but required in type 'DeepReadonlyObject<{ id: string; activate: (superdesk: DeepReadonlyObject<{ dataApi: IDataApi; dataApiByEntity: { article: { query(parameters: Pick<IQueryElasticParameters, "page" | "sort" | "filterValues" | "aggregations">): Promise<...>; }; }; ... 13 more ...; removeEventListener<T extends "articleEditStart" | "ar...'.
npm ERR! code ELIFECYCLE
npm ERR! errno 2
npm ERR! @ compile: `tsc`
npm ERR! Exit status 2
npm ERR! 
npm ERR! Failed at the @ compile script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2019-11-08T09_39_18_739Z-debug.log
child_process.js:660
    throw err;
    ^

Error: Command failed: cd /opt/superdesk/client/extensions/helloWorld2 && npm install && npm run compile
    at checkExecSyncError (child_process.js:621:11)
    at execSync (child_process.js:657:15)
    at /opt/superdesk/client/node_modules/superdesk-core/tasks/install-extensions.js:10:5
    at Array.forEach (<anonymous>)
    at Object.<anonymous> (/opt/superdesk/client/node_modules/superdesk-core/tasks/install-extensions.js:7:13)
    at Module._compile (internal/modules/cjs/loader.js:956:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:973:10)
    at Module.load (internal/modules/cjs/loader.js:812:32)
    at Function.Module._load (internal/modules/cjs/loader.js:724:14)
    at Function.Module.runMain (internal/modules/cjs/loader.js:1025:10) {
  status: 2,
  signal: null,
  output: [ null, null, null ],
  pid: 888,
  stdout: null,
  stderr: null
}
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! superdesk-core@1.31.2 setup-extensions: `node ./tasks/namespace-css-from-extensions.js && node ./tasks/install-extensions.js`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the superdesk-core@1.31.2 setup-extensions script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
npm WARN Local package.json exists, but node_modules missing, did you mean to install?

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2019-11-08T09_39_18_763Z-debug.log
npm WARN notsup Unsupported engine for karma@0.12.37: wanted: {"node":">=0.8 <=0.12 || >=1 <=2"} (current: {"node":"12.13.0","npm":"6.13.0"})
npm WARN notsup Not compatible with your version of node/npm: karma@0.12.37
npm WARN end-to-end-testing-helpers@1.0.3 requires a peer of protractor@5.4.2 but none is installed. You must install peer dependencies yourself.
npm WARN superdesk@ No description
npm WARN superdesk@ No repository field.
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@1.2.9 (node_modules/fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.2.9: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})

npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! superdesk-core@1.31.2 postinstall: `npm run setup-extensions && node ./tasks/compile-e2e-tests.js`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the superdesk-core@1.31.2 postinstall script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2019-11-08T09_39_23_115Z-debug.log
ERROR: Service 'superdesk' failed to build: The command '/bin/sh -c cd ./client && npm install && grunt build' returned a non-zero code: 1
```

**Comment by kottkrig** [comment]
It seems to be an issue with the extension `helloworld2`. After removal of that directory it builds and runs just fine.

**Comment by mehdignu** [comment]
`superdesk    | All services are started :3
superdesk    | 17:43:42 system | rest.1 started (pid=417)
superdesk    | 17:43:42 system | work.1 started (pid=420)
superdesk    | 17:43:42 system | beat.1 started (pid=423)
superdesk    | 17:43:42 system | wamp.1 started (pid=422)
superdesk    | 17:43:42 rest.1 | Traceback (most recent call last):
superdesk    | 17:43:42 rest.1 |   File "app.py", line 13, in <module>
superdesk    | 17:43:42 rest.1 |     import settings
superdesk    | 17:43:42 rest.1 |   File "/opt/superdesk/settings.py", line 14, in <module>
superdesk    | 17:43:42 rest.1 |     from superdesk.default_settings import INSTALLED_APPS, strtobool
superdesk    | 17:43:42 rest.1 | ModuleNotFoundError: No module named 'superdesk'
superdesk    | 17:43:42 system | rest.1 stopped (rc=1)
superdesk    | 17:43:42 system | sending SIGTERM to wamp.1 (pid 422)
superdesk    | 17:43:42 system | sending SIGTERM to work.1 (pid 420)
superdesk    | 17:43:42 system | sending SIGTERM to beat.1 (pid 423)
superdesk    | 17:43:42 system | wamp.1 stopped (rc=-15)
superdesk    | 17:43:42 system | beat.1 stopped (rc=-15)
superdesk    | 17:43:42 system | work.1 stopped (rc=-15)
superdesk exited with code 1`

any idea how to fix this issue ?

**Comment by nickdbush** [comment]
Can I check you're running ```docker-compose up``` in the ```docker``` directory?

**Comment by mehdignu** [comment]
@nickdbush thank you for your response ! this is what s running:

`(base) mehdis-macbook-pro-2:superdesk mehdi$  docker-compose -f docker/docker-compose-dev.yml up -d

mongodb is up-to-date
elastic is up-to-date
postfix is up-to-date
Starting kibana ... 
Starting kibana    ... done
Starting superdesk ... done`

and then superdesk fails and after it kibana

**Comment by mehdignu** [comment]
@nickdbush 
`superdesk exited with code 0
`

**Comment by mehdignu** [comment]
@nickdbush 
superdesk_—_docker-compose_◂_docker-compose_-f_docker_docker-compose-dev_yml_up_--build_—_90×30

**Comment by mehdignu** [comment]
@nickdbush  looks like the virtual env is not being activated

**Comment by nickdbush** [comment]
I'll look into it on Monday @mehdignu, until then can you try some of the other docker files? just ```docker-compose.yml``` (or ```docker-compose-prebuilt.yml```)?

File: docker/common.yml [hunk 2]
```diff
@@ -1,57 +1,69 @@
...
   container_name: mongodb
 
 redis:
-  image: redis:3
+  image: redis:5.0
   container_name: redis
 
 elastic:
-  image: elasticsearch:2
+  image: elasticsearch:7.4.0
```
Comment (petrjasek) [inline docker/common.yml]: this won't work now, only after merging some branches

File: docker/common.yml [hunk 3]
```diff
@@ -1,57 +1,69 @@
...
 
 grammalecte:
   build: grammalecte/
   container_name: grammalecte
 
+# logstash:
+#   build: logstash/
+#   container_name: logstash
+
+kibana:
```
Comment (petrjasek) [inline docker/common.yml]: I would disable this one too, I think it was supposed to be the ui for logstash

File: docker/common.yml [hunk 4]
```diff
@@ -1,57 +1,69 @@
...
+    - GRAMMALECTE_URL=http://grammalecte:9999
+    - MONGO_URI=mongodb://mongodb/test
+    - PUBLICAPI_MONGO_URI=mongodb://mongodb/test_pa
+    - LEGAL_ARCHIVE_URI=mongodb://mongodb/test_la
+    - ARCHIVED_URI=mongodb://mongodb/test_ar
+    - ELASTICSEARCH_URL=http://elastic:9200
+    - ELASTICSEARCH_INDEX=superdesk
+    - CELERY_BROKER_URL=redis://redis:6379/1
+    - REDIS_URL=redis://redis:6379/1
+    - LOG_SERVER_ADDRESS=logstash
```
Comment (petrjasek) [inline docker/common.yml]: these `LOG_*` vars shouldn't be set probably if logstash is disabled

File: server/settings.py [hunk 5]
```diff
@@ -61,8 +61,8 @@ def env(variable, fallback_value=None):
...
-ANALYTICS_ENABLE_SCHEDULED_REPORTS = strtobool(
-    env('ANALYTICS_ENABLE_SCHEDULED_REPORTS', 'true')
-)
-HIGHCHARTS_SERVER_HOST = env('HIGHCHARTS_SERVER_HOST', 'localhost')
-HIGHCHARTS_SERVER_PORT = env('HIGHCHARTS_SERVER_PORT', '6060')
+# ANALYTICS_ENABLE_SCHEDULED_REPORTS = strtobool(
+#     env('ANALYTICS_ENABLE_SCHEDULED_REPORTS', 'true')
+# )
+# HIGHCHARTS_SERVER_HOST = env('HIGHCHARTS_SERVER_HOST', 'localhost')
+# HIGHCHARTS_SERVER_PORT = env('HIGHCHARTS_SERVER_PORT', '6060')
```
Comment (petrjasek) [inline server/settings.py]: you could keep the settings but set the env vars to be empty/false via docker

**Comment by nickdbush** [comment]
> 
> 
> It seems to be an issue with the extension `helloworld2`. After removal of that directory it builds and runs just fine.

Merged with upstream master which resolved this issue

File: server/settings.py (hunk 5 above)
Comment (nickdbush) [reply server/settings.py]: Leaving these lines but without env variables from docker causes a ```ValueError```, as an empty string is casted to an integer

**Comment by nickdbush** [comment]
@mehdignu if you follow the new instructions in the readme does it fix your earlier issues?

**Comment by mehdignu** [comment]
@nickdbush looks great ! but i can not connect on the DOCKER_IP (for me it's 192.168.99.100)

**Comment by nickdbush** [comment]
> 
> 
> @nickdbush looks great ! but i can not connect on the DOCKER_IP (for me it's 192.168.99.100)

Have you manually set that environment variable?

**Comment by mehdignu** [comment]
@nickdbush 
docker_—_-bash_—_90×30

**Comment by sagun786** [comment]
After I run the command
./scripts/docker-local-demo.sh

Command python setup.py egg_info failed with error code 1 in /tmp/pip_build_root/python-twi
tter
Storing debug log for failure in /root/.pip/pip.log
ERROR: Service 'superdesk' failed to build: The command '/bin/sh -c pip3 install -U -r requ
irements.txt' returned a non-zero code: 1

Is there an updated flow of docker installation (where the image has all the above resolved)

**Pull request closed without merging**